src/sonic-host-services/scripts/hostcfgd

```diff
@@ -57,15 +71,210 @@ def obfuscate(data):
+            self._cached_config[feature_name].auto_restart = feature.auto_restart
+
+        # Enable/disable the container service if the feature state was changed from its previous state.
+        if self._cached_config[feature_name].state != feature.state:
+            if self.update_feature_state(feature):
+                self._cached_config[feature_name].state = feature.state
+            else:
+                self.resync_feature_state(self._cached_config[feature_name])
+
+    def update_all_features_config(self):
+        feature_table = self._config_db.get_table('FEATURE')
+        for feature_name in feature_table.keys():
+            if not feature_name:
+                syslog.syslog(syslog.LOG_WARNING, "Feature is None")
+                continue
+
+            feature = Feature(feature_name, feature_table[feature_name], self._device_config)
+            self._cached_config.setdefault(feature_name, feature)
+
+            self.update_feature_auto_restart(feature)
+
+            self.update_feature_state(feature)
+            self.resync_feature_state(feature)
+
+    def update_feature_state(self, feature):
+        cached_feature = self._cached_config[feature.name]
+        enable = False
+        disable = False
+
+        # Allowed transitions:
+        #  None           -> always_enabled
+        #                 -> always_disabled
+        #                 -> enabled
+        #                 -> disabled
+        #  always_enabled -> always_disabled
+        #  enabled        -> disabled
+        #  disabled       -> enabled
+        if cached_feature.state is None:
+            enable = feature.state in ("always_enabled", "enabled")
+            disable = feature.state in ("always_disabled", "disabled")
+        elif cached_feature.state in ("always_enabled", "always_disabled"):
+            disable = feature.state == "always_disabled"
+        elif cached_feature.state in ("enabled", "disabled"):
+            enable = feature.state == "enabled"
+            disable = feature.state == "disabled"
+        else:
+            syslog.syslog(syslog.LOG_INFO, "Feature {} service is {}".format(feature.name, cached_feature.state))
+            return False
+
+        if not enable and not disable:
+            syslog.syslog(syslog.LOG_ERR, "Unexpected state value '{}' for feature {}"
+                          .format(feature.state, feature.name))
+            return False
+
+        if enable:
+            self.enable_feature(feature)
+            syslog.syslog(syslog.LOG_INFO, "Feature {} is enabled and started".format(feature.name))
+
+        if disable:
+            self.disable_feature(feature)
+            syslog.syslog(syslog.LOG_INFO, "Feature {} is stopped and disabled".format(feature.name))
+
+        return True
+
+    def update_feature_auto_restart(self, feature):
+        restart_config = "always" if feature.auto_restart == "enabled" else "no"
+        service_conf = "[Service]\nRestart={}\n".format(restart_config)
```

Comment:
Do we consider put the service configuration entry `RestartSec=30` in the `service_conf` since `RestartSec` is usually configured with `Restart`? Please review this: https://www.freedesktop.org/software/systemd/man/systemd.service.html#RestartSec=

Reply:
We have default RestartSec=30 in all service files and there is no way to override it via FEATURE table configuration, so it is not put in service_conf.

Reply:
What I suggested is not to override the `RestartSec=30` configuration. Would it be better to do as following:

`service_conf = "[Service]\nRestart={}\nRestartSec=30\n".format(restart_config)`

Reply:
But then user will not be able to set this value on a per service basis (like changing systemd serivce file) - he must change this hardcoded value in hostcfgd but then it will apply to all services.
I would like to wait till we introduce this configuration via FEATURE table.

Reply:
I am just wondering if we want to provide the user the ability to manually configure the `RestartSec=` in each service configuration file `like changing systemd service file`, do we need also provide such ability for users to manually change the `Restart=` value `like changing systemd service file`? The user may want to change `Restart=always` to `Restart=on-failure` or `Restart=on-abnormal`?

Reply:
This means not a regular user, but developer. How to change restart sec for a particular service at build time in service file if it is hardcoded to be 30 sec for all of them? I would like not to introduce such limitation.
A restart configuration change is provided to the user by "config feature auto-restart". A more granular restart options are not exposed by SONiC.

Reply:
I agreed with you we should not introduce the limitation to hardcode the `RestartSec=30`.

I know the user can use the CLI `sudo config feature autorestart <container_name> <disabled |
enabled>` to disable or enable the auto-restart feature of each container. 

What I am thinking is  whether it is a good way to hardcode `Restart=always` if the developer wants to introduce `Restart=on-failure` or `Restart=on-abnormal` in future?

Reply:
It is different for Restart= because it is exposed to user with "auto_restart" field in SONiC. Lets say today user knows he can configure auto restart via config db, the container auto restarts type is "always". If someone changes it to "on-abnormal" in next version - that will mean that despite auto restart is enabled in config db service will not restart in case service exited with failure. This will be confusing, if developer wants to introduce these restart options it should now be done through CONFIG DB and auto_restart will have more faules e.g: - <enabled|disabled|on-failure|on-abnormal>.

Reply:
@yozhao101 Kindly reminder

Reply:
Thanks so much for your response. This discussion is resolved.